implementations/ariadne/server.py

```diff
@@ -2,137 +2,216 @@
-    return {'email': 'support@apollographql.com',
-            'totalProductsCreated': 1337}
+    return user_data
 
 
 @product.reference_resolver
 def resolve_product_reference(_, _info, representation):
     if "sku" in representation and "package" in representation:
-        return get_product_by_sku_and_package(representation['sku'], representation['package'])
+        return get_product_by_sku_and_package(
+            representation["sku"], representation["package"]
+        )
     if "sku" in representation and "variation" in representation:
-        return get_product_by_sku_and_variation(representation['sku'], representation['variation'])
-    return get_product_by_id(representation['id'])
+        return get_product_by_sku_and_variation(
+            representation["sku"], representation["variation"]
+        )
+    return get_product_by_id(representation["id"])
 
 
-schema = make_federated_schema(type_defs, [query, product])
-application = GraphQL(schema)
+# ------- deprecated product -------
 
-products = [
-    {
-        "id": "apollo-federation",
-        "sku": "federation",
-        "package": "@apollo/federation",
-        "variation": "OSS",
-    },
-    {
-        "id": "apollo-studio",
-        "sku": "studio",
-        "package": "",
-        "variation": "platform",
-    },
-]
 
+@deprecated_product.reference_resolver
+def resolve_deprecated_product_reference(_, _info, representation):
+    return deprecated_product_data
 
-def get_product_variation(reference):
-    if reference['variation']:
-        return {'id': reference['variation']}
-    variation = next((product for product in products if product['id']
-                     == reference['id']), None)
-    return {'id': variation}
 
+# ------- product research -------
 
-def get_product_by_id(id):
-    return next((product for product in products if product['id']
-                == id), None)
 
+@product_research.reference_resolver
+def resolve_product_research_reference(_, _info, representation):
+    case_number = representation["study"]["caseNumber"]
 
-def get_product_by_sku_and_package(sku: str, package: str) -> Optional[dict]:
-    data = next(
+    return next(
         (
-            product
-            for product in products
-            if product["sku"] == sku and product["package"] == package
+            research
+            for research in products_research_data
+            if research["study"]["caseNumber"] == case_number
         ),
         None,
     )
 
-    if not data:
-        return None
 
-    return data
+# ------- user -------
 
 
-def get_product_by_sku_and_variation(sku: str, variation: dict) -> Optional[dict]:
-    data = next(
-        (
-            product
-            for product in products
-            if product["sku"] == sku and product["variation"] == variation["id"]
-        ),
-        None,
-    )
+@user.reference_resolver
+def resolve_user_reference(_, _info, representation):
+    return user_data
```

Comment:
while this will pass the tests I believe you should be resolving it based on the matching email